ERK optional forward parameter sensitivities `S_p` (end-to-end)

ivandoornen opened the pull request:
## Summary

This PR adds optional forward parameter sensitivities \(S_p\) to the simulation layer, with ERK support for propagating them, exposes per-stage parameter sensitivities at the OCP level, and makes the feature usable from the MATLAB interface and generated code.

For an integrator step
$x_{n+1} = f(x_n, u_n, p_n),$
we define the forward parameter sensitivities
$S_p = \frac{\partial x_{n+1}}{\partial p}, \quad S_p \in \mathbb{R}^{n_x \times n_{p}},$


---

## Changes in this PR

### Common sim API (`acados/sim/sim_common.[ch]`)
- Add optional parameter inputs to `sim_in`:
  - `p` (stagewise, length `np`)
- Add `S_p` to `sim_out` with layout `[nx x np]` (column-major).
- Add `sens_forw_p` option to `sim_opts` (default: `false`).

### ERK integrator (`acados/sim/sim_erk_integrator.[ch]`)
- Add `np` to ERK dims.
- Add a new external function hook: `expl_vde_forw_p` for parameter-sensitivity propagation.
- When `sens_forw && sens_forw_p && np> 0`:
  - propagate `S_p` across stages via `expl_vde_forw_p`
  - write results to `sim_out->S_p`

### IRK / lifted-IRK / GNSF compatibility
- Update dims/plumbing in `sim_irk_integrator`, `sim_lifted_irk_integrator`, and `sim_gnsf` so they remain compatible with the extended sim API.
- These integrators do not compute `S_p` yet.

### OCP dynamics (`acados/ocp_nlp/ocp_nlp_dynamics_cont.c/h`)
- Extend the continuous dynamics memory with a BLASFEO matrix `S_p` of size `[nx_{k+1} × np]`.
- Implement `dims_get(..., "np", ...)` to return `np` based on the underlying sim dims (`np`).
- In the dynamics evaluation callback:
  - If parameter sensitivities are enabled in the simulator options, pack `sim_out->S_p` into the cached BLASFEO `S_p`.
  - Otherwise, zero `S_p` to keep memory contents deterministic.
- Provide `memory_get(..., "S_p", ...)` to export the cached `S_p(k)` to a dense buffer.

### C interface (`interfaces/acados_c/ocp_nlp_interface.c`)
- Add a dimension query for `"S_p"`:
  - `ocp_nlp_qp_dims_get_from_attr(..., "S_p", dims_out) -> [nx_{k+1}, np]`
- Add a getter for `"S_p"`:
  - `ocp_nlp_get_at_stage(solver, k, "S_p", S_p_out)` for stages `k = 0..N-1`
  - Terminal stage is handled explicitly (warning / empty).

---

## MATLAB test

### examples/acados_matlab_octave/test/test_sens_forw_p.m (new)
- Adds a finite-difference validation test for ERK parameter sensitivities:
  - enables `sens_forw_p='true'` and compares `S_p` against finite differences
  - also checks `Sx` / `Su` FD consistency

---

## MATLAB interface plumbing

### interfaces/acados_matlab_octave/AcadosOcpOptions.m
- Adds option property `sens_forw_p` (default `false`) so OCP creation can request param sensitivities in dynamics.

### interfaces/acados_matlab_octave/AcadosSimOptions.m
- Adds `sens_forw_p` (default `false`).

### interfaces/acados_matlab_octave/AcadosSim.m
- Extends option validation: `opts.sens_forw_p` must be logical.

### interfaces/acados_matlab_octave/acados_sim_opts.m (legacy sim interface)
- Adds legacy string option `sens_forw_p`.

### interfaces/acados_matlab_octave/setup_AcadosSim_from_legacy_sim_description.m
- When upgrading legacy sim descriptions, forwards `sens_forw_p` if present (else sets to `false`).

---

## Code generation (CasADi) for param-direction forward VDE

### interfaces/acados_matlab_octave/generate_c_code_explicit_ode.m
- Selects the effective parameter symbol:
  - use `model.p` if present, else an empty CasADi vector (typed SX/MX).
- If `np > 0`, creates a new CasADi function:
  - `{{model.name}}_expl_vde_forw_p`
  - computes `vdeP = f_p + A * Sp` (with `A = ∂f/∂x`, `f_p = ∂f/∂p`)
  - signature uses inputs `{x, Sp, u, p}` and outputs `{vdeP}`.
  - This is the function the ERK integrator calls when `sens_forw_p` is enabled.

---

## MEX additions for access to S_p and the new external function

### interfaces/acados_matlab_octave/sim_create_ext_fun.c
- Extends the external function struct with `dyn_expl_vde_forw_p` pointer storage.

### interfaces/acados_matlab_octave/sim_get.c
- Reads `np` from `sim_dims`.
- Adds `sim.get('S_p')` returning an `[nx x np]` MATLAB matrix.

### interfaces/acados_matlab_octave/ocp_get.c
- Adds `ocp.get('S_p')`:
  - `ocp.get('S_p')` returns a `1xN` cell array, each cell `[nx_{k+1} x np]`
  - `ocp.get('S_p', k)` returns a single `[nx_{k+1} x np]` matrix
- Layout note: returned matrices are MATLAB-native (column-major, leading dimension `nx_{k+1}`).

---

## Template plumbing (acados_template / Tera)

### interfaces/acados_template/.../model.in.h
- Declares prototypes for `*_expl_vde_forw_p` (guarded by `np>0`).

### interfaces/acados_template/.../acados_sim_solver.in.h / .in.c
- Adds a capsule pointer `sim_expl_vde_forw_p`.
- Allocates/maps `sim_expl_vde_forw_p` when parameters exist.
- Registers external function with the sim model via `model_set("expl_vde_forw_p", ...)`.
- Sets sim option `"sens_forw_p"` only if user enabled it and `np>0`.

### interfaces/acados_template/.../acados_solver.in.h / .in.c
- Adds `expl_vde_forw_p` array (ERK stages) when parameters exist.
- Registers it into the dynamics via `ocp_nlp_dynamics_model_set_external_param_fun(...,"expl_vde_forw_p",...)`.
- Sets solver option `"dynamics_sens_forw_p"` at all stages when enabled (and parameters exist).
- Frees the allocated `expl_vde_forw_p` array during teardown.

---

## Behavior / compatibility

- Off by default: `sens_forw_p = false`.
- `S_p` is only propagated when `sens_forw && sens_forw_p && np> 0`.
- The additional `S_p` workspace in ERK is only used when the above condition holds.
- If `np = 0` or `sens_forw_p = false` , existing behavior is unchanged.

External function requirement:
- If `sens_forw_p` is enabled (and `sens_forw` is enabled) with `np> 0`, the ERK requires an externally supplied `expl_vde_forw_p`.

OCP exposure:
- If sensitivities are enabled, `S_p(k)` is cached in the OCP dynamics memory and can be retrieved via the new `"S_p"` getter.
- If sensitivities are not enabled, `S_p(k)` is returned as zero (deterministic memory contents).
- `S_p_out` is returned as a dense matrix with leading dimension `nx_{k+1}` (i.e. column-major).

Backward compatibility:
- Default is unchanged (`sens_forw_p = false` everywhere).
- No additional memory or external functions are used unless `sens_forw_p` is explicitly enabled.
- When parameters are absent, generated code auto-disables `sens_forw_p` even if requested.

---

## Follow-up plan (roles + dependencies)

1. zoRO / Simulink integration
   - Wire `S_p` into the zoRO workflow (tightening/backoff terms that depend on parameter uncertainty).
   - Update custom update templates and Simulink S-function integration / payload streaming as needed.

FreyJo commented on acados/sim/sim_common.c:
```diff
@@ -80,14 +80,19 @@ acados_size_t sim_in_calculate_size(void *config_, void *dims)
     acados_size_t size = sizeof(sim_in);
 
-    int nx, nu, nz;
+    int nx, nu, nz, np = 0, np_global = 0;
```
Initialization of to 0 not necessary as getter will overwrite it later anyway.
Comments regarding "optional" can be removed, as also "nu" and "nz" can be 0.

ivandoornen commented:
Hi,  
Thank you for reviewing my code!  

> Somehow, I cannot make the CI tests run on the PR as usual. Did you run them? Otherwise I can try some more, probably pushing as a maintainer of acados helps.

I was also surprised it did not run the CI tests, like it did with my last (now closed) PRs, even though they normally seem to be triggered upon opening a pull request. I thought perhaps a maintainer needed to trigger them, but apparently that is not the case either. I did run `test_all_examples.m` locally, but I do not think I can run or trigger the rest of the CI myself.

> I think that `p_eff` introduces some ambiguity and we should rather just support either p or p_global and later think about how to generalize it to the other one.

I see your point that `p_eff` might be a bit ambiguous. I could either change it to support only `p` or only `p_global`, or define a separate `sens_forw_p` and `sens_forw_p_global` so it is explicit which one is used.

> In particular, the function that you need to call (`expl_vde_for_p`) needs to be defined in the interfaces to evaluate the derivative of some specific parameter. Also, this definition should be part of this PR.

I split the series of PRs up into 4, with the first being the sim side, the second the OCP side, the third the MATLAB interfaces and templates, and the last the zoRO and Simulink implementation. In the third PR I modified `generate_c_code_explicit_ode.m` to generate `expl_vde_for_p`. The other 3 PRs are already visible on my fork.

> Do you have a test for evaluating parametric sensitivities of ERK? I think it would be great to add one here.

In the third PR I added a test script that compares the parametric sensitivities of ERK against a finite difference approach for `pendulum_on_cart_model_with_param()` (similar to the existing `test_sens_*m` scripts). Since the test requires the MATLAB interface, I only added it to the third PR.

FreyJo commented:
Yes, acados does support `p` and `p_global` at the same time.
However, `p_global` is added as an input to the function definitions later in the `GenerateContext`.
Another limitation of `p_global` is that they have not been added to the integrator (yet).
Thus, sensitivities with respect to `p_global` cannot be tested with the standalone integrator interface.

FreyJo commented:
I think it would be best to implement both sensitivities wrt p and p_global separately in ERK to be able to test both.
Alternatively, we need to add a similar test through the OCP solver interface.

Regarding the PR structure: I suggest also adding the changes from your PRs 2 and 3 directly.
This allows to actually test the new functionality when its added and also simplify the work for you.
Right now, I would need to test a checkout a different branch in order to test the added feature.

ivandoornen commented:
> Regarding the PR structure: I suggest also adding the changes from your PRs 2 and 3 directly. This allows to actually test the new functionality when its added and also simplify the work for you. Right now, I would need to test a checkout a different branch in order to test the added feature.

Just to confirm: would you prefer me to open PR2 and PR3 here as stacked PRs on top of PR1, or squash them into a single PR so everything needed for testing is in one place?

FreyJo commented:
> For now I’m leaning towards supporting S_p for p only, and putting all parameters I want sensitivities for into p, while keeping the rest in p_global.

This is fine for me!
Indeed extending integrator to use `p_global` is a bit more involved.
The parameters (or variants of them) are set in the general solver wrapper.
So, yes, easier if you just do p, also for testing.

I meant squashing them into one PR, so you can just apply the commits on the branch of this PR.

ivandoornen commented:
> I meant squashing them into one PR, so you can just apply the commits on the branch of this PR.

I have now squashed them into this PR. It has become a bit messy, so if you would like I could also create a fresh PR.

FreyJo commented on acados/ocp_nlp/ocp_nlp_dynamics_cont.c:
```diff
@@ -379,6 +394,13 @@ acados_size_t ocp_nlp_dynamics_cont_memory_calculate_size(void *config_, void *d
     size +=
         config->sim_solver->memory_calculate_size(config->sim_solver, dims->sim, opts->sim_solver);
 
+    // buffer for parameter sensitivities S_p (nx1 x np_eff)
```
Please replace `np_eff` with `np` everywhere.

FreyJo commented on acados/ocp_nlp/ocp_nlp_dynamics_cont.c:
```diff
@@ -577,6 +609,18 @@ void ocp_nlp_dynamics_cont_memory_get(void *config_, void *dims_, void *mem_, co
+		config->sim_solver->dims_get(config->sim_solver, dims->sim, "np", &np);
+		if (np > 0)
+		{
+			// layout: row-major double[nx1 * np_eff]
```
I think it is column-major, also compare your comment below.

FreyJo commented on acados/sim/sim_common.c:
```diff
@@ -275,16 +289,18 @@ acados_size_t sim_out_calculate_size(void *config_, void *dims)
 
     size += nx * sizeof(double);                // xn
     size += nx * NF * sizeof(double);           // S_forw
+    size += nx * np * sizeof(double);       // S_p (parameter sens, may be 0)
```
Most new memory assignments and all potentially costly operations should be conditional on the option: `sens_forw_p`.
In particular, since often np >> nx, assigning nx * np memory can increase the memory footprint significantly.

Here in `sim_out`, there is the issue that the assign functions do not depend on `opts`.
I would mitigate this, by not including `S_p` in `sim_out` and instead directly get it from the integrator memory `sim_erk_memory_get`.

FreyJo commented on acados/sim/sim_erk_integrator.c:
```diff
@@ -337,6 +356,17 @@ void *sim_erk_memory_assign(void *config, void *dims, void *opts_, void *raw_mem
+    sim_erk_dims *erk_dims = (sim_erk_dims *) dims;
+    sim_opts *opts = (sim_opts *) opts_;
+
+    if (opts->sens_forw && opts->sens_forw_p )
```
why do you check for both? I would have just checked `opts->sens_forw_p`.

FreyJo commented on interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:
```diff
@@ -2471,6 +2481,16 @@ static void {{ model.name }}_acados_create_set_opts({{ model.name }}_solver_caps
     free(sim_method_jac_reuse);
   {%- endif %}
 
+{%- if solver_options.integrator_type == "ERK" %}
```
I would remove the if here and add sanity checks that `sens_forw_p` can only be True if ERK is used in `make_consistent` in the high level interfaces.

FreyJo reviewed (commented):
Looks good!
Can you answer the remaining comments?

ivandoornen commented:
> Ok, thanks for the explanation. I would then rather replace the corresponding `if (opts->sens_forw)` checks with `if (opts->sens_forw || opts->sens_forw_p)`. Such that S p is always computed when `sens_forw_p` is true. And potentially S x is computed despite `sens_forw = False`, just as a by product. Does that make sense?

I see, so essentially you would prefer that whenever `opts->sens_forw_p` you take the VDE path, regardless of  `opts->sens_forw`? Makes sense to me!
